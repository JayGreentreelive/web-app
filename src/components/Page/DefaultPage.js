import React from 'react'
import PropTypes from 'prop-types'
import {
  useQuery
} from 'react-apollo'
import {
  mapEdgesToNodes,
} from '../../utils'
import getWebPageBlockItems from '../../queries/getWebPageBlockItems'

import SEO from '../../seo'
import PixelManager from '../PixelManager'

import { Loader } from '@christfellowshipchurch/web-ui-kit'
import ContentBlock from '../ui/ContentBlock'
import GroupBlock from '../ui/GroupBlock'


const DefaultPage = ({ title, match: { params: { page } } }) => {
  PixelManager.initWithPageView(`/${page || ''}`)

  const website = process.env.REACT_APP_WEBSITE_KEY
  const { loading, error, data } = useQuery(getWebPageBlockItems, { variables: { website, title: page || title } })

  if (loading) return (
    <div className="vh-100 vw-100 d-flex justify-content-center align-items-center bg-light">
      <Loader />
    </div>
  )

  if (error) {
    console.error("ERROR: ", error)
    return <h1 className="text-center">There was an error loading the page. Please try again.</h1>
  }

  const blockItems = mapEdgesToNodes(data.getWebsitePageContentByTitle.childContentItemsConnection)
  const {
    metaDescription,
    metaKeywords,
    openGraphProtocols,
    twitterProtocols,
  } = blockItems

  return (
    <>
      <SEO
        title={`${page || title} - Christ Fellowship Church`}
        description={metaDescription}
        keywords={metaKeywords}
        openGraphProtocols={openGraphProtocols}
        twitterProtocols={twitterProtocols}
      />

      {blockItems.map((item, i) => {
        const classenames = i % 2 === 0 ? 'bg-white' : 'bg-light'
        switch (item.__typename) {
          case 'WebsiteBlockItem':
            return (
              <div className={classenames} key={i}>
                <ContentBlock {...item} />
              </div>
            )
          case 'WebsiteGroupItem':
            return (
              <div className={classenames} key={i}>
                <GroupBlock {...item} />
              </div>
            )
          default:
            return <h1 className="text-center" key={i}>{item.title}</h1>
        }
      })}
    </>
  )
}

DefaultPage.defaultProps = {
  title: 'home',
}

DefaultPage.propTypes = {
  title: PropTypes.string,
  match: PropTypes.shape({
    params: PropTypes.shape({
      page: PropTypes.string
    })
  })
}

export default DefaultPage
