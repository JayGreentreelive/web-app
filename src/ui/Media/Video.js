import React, { createRef, useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames'
import Hls from 'hls.js';
import { isMobile, isIOS } from 'react-device-detect'

import { Icon } from '../Icons';

const MediaVideo = ({ source, poster, isLive, showControls, playIcon }) => {
  const [showPlayButton, setShowPlayButton] = useState(showControls);
  const [showMuteButton, setShowMuteButton] = useState(isMobile && isLive);



  let videoProps = showControls
    ? {
      playsInline: false,
      autoPlay: false,
      loop: false,
      muted: false,
      controls: !showPlayButton,
    }
    : {};

  if (isLive || source.includes('m3u8') && !isMobile) {
    videoProps = {
      ...videoProps,
      autoPlay: true,
      playsInline: true
    }
  }

  if(isMobile){
    videoProps = {
      ...videoProps,
      muted: isLive,
      playsInline: true,
      showControls: false,
      autoPlay: isLive
    }
  }

  const videoRef = createRef();

  const createHLSurl = () => {
    let hls = new Hls({});
    hls.loadSource(source);
    hls.attachMedia(videoRef.current);
    hls.on(Hls.Events.MANIFEST_PARSED, () => {
      videoRef.current.play()
    })
  }

  const muteButtonClick = () => {
    setShowMuteButton(false)
    videoRef.current.muted = false
  }

  const playButtonClick = () => {
    if(source.includes('m3u8')){
      createHLSurl()
    }
    else{
      videoRef.current.play();
    }
    setShowPlayButton(false);
  };

  useEffect(() => {
    if (isLive || isIOS) return playButtonClick()
  }, [videoRef])

  return (
    <div>
      <video
        {...videoProps} 
        poster={poster}
        ref={videoRef}
        controlsList="nodownload"
        style={{
          objectFit: 'cover'
        }}
      >
        <source  
          type='video/mp4' 
          src={source}
        />
      </video>
      {showPlayButton &&
        <div className="fill d-flex justify-content-center align-items-center" style={{ zIndex: 1000 }}>
              <button
                className="btn btn-icon"
                onClick={playButtonClick}
              >
                <Icon
                  name='play-circle' 
                  size={playIcon.size} 
                  fill={playIcon.color}
                />
              </button>
        </div>
      }
      {showMuteButton &&
        <div
          className="fill d-flex justify-content-center align-items-center" 
          style={{ zIndex: 1000 }}
        >
          <h4
            className={classnames(
              'card',
              'bg-dark',
              'p-1',
              'text-white'
            )}
            onClick={muteButtonClick}
          >
            Tap to Unmute
          </h4>
          {/* <button
            className='btn btn-icon p-1'
            onClick={muteButtonClick}
          >
            <Icon
              name={'ban'} 
              size={playIcon.size}
              fill={playIcon.color}
            />
          </button> */}
        </div>
      }
    </div>
  )};

// Default state of the video is to be a silent,
//  inline, looped video to be used like a background video
const defaultProps = {
  className: '',
  playsInline: true,
  autoPlay: true,
  loop: true,
  muted: true,
};

const propTypes = {
  className: PropTypes.string,
  source: PropTypes.string.isRequired,
  playsInline: PropTypes.bool,
  autoPlay: PropTypes.bool,
  loop: PropTypes.bool,
  muted: PropTypes.bool,
};

MediaVideo.defaultProps = defaultProps;
MediaVideo.propTypes = propTypes;

export default MediaVideo;
