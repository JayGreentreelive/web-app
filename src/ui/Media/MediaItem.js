import React, { createRef, useState } from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import { keys } from 'lodash'
import { faPlayCircle } from '@fortawesome/fontawesome-pro-light'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'

import Image from './Image'
import Video from './Video'

const MediaItem = ({
  ratio,
  imageUrl,
  imageAlt,
  videoUrl,
  className,
  children,
  rounded,
  circle,
  showControls,
  playIcon,
  overlay,
  gradient,
  gradientDirection,
  withHover,
  style
}) => {
  const showVideoControls = showControls && !children
  const [showPlayButton, setShowPlayButton] = useState(showVideoControls)
  const videoProps = showVideoControls
    ? {
      playsInline: false,
      autoPlay: false,
      loop: false,
      muted: false,
      controls: !showPlayButton,
    }
    : {}
  let videoRef = createRef()
  const playButtonClick = () => {
    videoRef.current.play()
    setShowPlayButton(false)
  }
  let ratioClass = typeof ratio === 'string'
    ? `embed-responsive-${ratio}`
    : keys(ratio).map(n => `embed-responsive-${n}-${ratio[n]}`.replace('-xs', ''))

  if (circle) {
    ratioClass = '1by1'
  }

  // TODO : test where the showControls is passed in, but no value URL exists

  return (
    <div
      className={classnames(
        className,
        'embed-responsive',
        ratioClass,
        {
          'rounded': rounded && !circle,
          'rounded-circle': circle,
          'scale-media-up-on-hover': withHover
        }
      )}
      style={style}
    >
      <Image
        source={imageUrl}
        alt={imageAlt}
        className={classnames(
          'embed-responsive-item',
        )}
      />

      {videoUrl &&
        <Video
          className={classnames(
            'embed-responsive-item',
          )}
          source={videoUrl}
          {...videoProps}
          ref={videoRef}
          poster={imageUrl}
        />
      }

      {/* TODO : add gradient abilities */}
      {(gradient || overlay) &&
        <div
          className={classnames(
            'w-100',
            'h-100',
            'absolute-center',
            'opacity-65',
            {
              [`bg-${overlay}`]: !!overlay,
              [`gradient-${gradient}-${gradientDirection}`]: !!gradient
            }
          )}
        ></div>
      }

      {
        (children || (showPlayButton && videoUrl)) &&
        <div className='fill d-flex justify-content-center align-items-center'>
          {(showVideoControls && videoRef)
            ? (
              <button
                className="btn btn-icon"
                onClick={playButtonClick} >
                <FontAwesomeIcon icon={faPlayCircle} size={playIcon.size} color={playIcon.color} />
              </button>
            )
            : children}
        </div>
      }

    </div >
  )
};

const defaultProps = {
  ratio: '1by1',
  videoUrl: null,
  className: '',
  showControls: false,
  playIcon: {
    as: null,
    color: 'white',
    size: '3x',
  },
  overlay: null,
  gradient: null,
  gradientDirection: 'bottom-top',
  withHover: false,
}

const RATIOS = ['1by1', '4by3', '16by9', '21by9']
const propTypes = {
  ratio: PropTypes.oneOfType([
    PropTypes.oneOf(RATIOS),
    PropTypes.shape({
      xs: PropTypes.oneOf(RATIOS),
      sm: PropTypes.oneOf(RATIOS),
      md: PropTypes.oneOf(RATIOS),
      lg: PropTypes.oneOf(RATIOS),
      xl: PropTypes.oneOf(RATIOS)
    })
  ]),
  imageUrl: PropTypes.string.isRequired,
  imageAlt: PropTypes.string.isRequired,
  videoUrl: PropTypes.string,
  className: PropTypes.string,
  showControls: PropTypes.bool,
  withHover: PropTypes.bool,
  playIcon: PropTypes.shape({
    as: PropTypes.element, // TODO : add support
    color: PropTypes.string,
    size: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number
    ]),
  }),
  overlay: PropTypes.oneOf([
    "primary",
    "secondary",
    "success",
    "info",
    "warning",
    "danger",
    "light",
    "dark",
    "black",
  ]),
  gradient: PropTypes.oneOf([
    "primary",
    "secondary",
    "success",
    "info",
    "warning",
    "danger",
    "light",
    "dark",
    "black",
  ]),
  gradientDirection: PropTypes.oneOf([
    'bottom-top'
  ])
}

MediaItem.defaultProps = defaultProps;
MediaItem.propTypes = propTypes;

export default MediaItem;
