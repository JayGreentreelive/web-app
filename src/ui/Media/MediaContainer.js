import React, { createRef, useState } from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import { keys } from 'lodash'
import { faPlayCircle } from '@fortawesome/fontawesome-pro-light'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'

import MediaItem from './MediaItem'

const MediaContainer = ({
  children,
  style,
  className,
  ...mediaItemProps
}) => {
  if (!children || children.length === 0) return <MediaItem
    {...mediaItemProps}
    className={className}
    style={style}
  />

  return <div
    style={style}
    className={classnames(
      'overflow-hidden',
      'p-relative',
      className
    )}
  >
    <MediaItem
      {...mediaItemProps}
      className="fill"
      style={{ zIndex: -100, }}
    />
    {children}
  </div>
}

MediaContainer.defaultProps = {
  style: {},
  className: '',
}

MediaContainer.propTypes = {
  style: PropTypes.object,
  className: PropTypes.string,
}

export default MediaContainer
