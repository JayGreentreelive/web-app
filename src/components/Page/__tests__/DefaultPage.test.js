import React from "react"
import {
  Accordion,
  Carousel,
  Row,
  ,
  Block,
  Button
} from "@christfellowshipchurch/web-ui-kit"
import Media from '../../../ui/Media'
import { ContentBlock, GroupBlock } from '../DefaultPage'
import renderer from "react-test-renderer"

describe("DefaultPage", () => {

  //Content Blocks
  it("renders ContentBlock with background layout with Media in background", () => {
    const tree = renderer.create(
      <Media
        imageUrl=''
        imageAlt=''
      >
        <Block>
        </Block>
      </Media>
    );
    expect(tree).toMatchSnapshot();
  });
  it("renders ContentBlock without background layout", () => {
    const tree = renderer.create(
      <Block>
        <Block.Subtitle />
        <Block.Title />
        <Block.Body />
      </Block>
    );
    expect(tree).toMatchSnapshot();
  });


  //BlockItem and GroupItem
  it("renders single Block", () => {
    const tree = renderer.create(
      <div>
        <ContentBlock />
      </div>
    );
    expect(tree).toMatchSnapshot();
  });
  it("renders a groupItem", () => {
    const tree = renderer.create(
      <div>
        <GroupBlock />
      </div>
    );
    expect(tree).toMatchSnapshot();
  });
});
