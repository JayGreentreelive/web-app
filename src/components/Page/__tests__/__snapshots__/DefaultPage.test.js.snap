// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultPage renders ContentBlock with background layout with Media in background 1`] = `
<div
  className="embed-responsive embed-responsive-1by1 "
>
  <div
    className="fill d-flex justify-content-center align-items-center"
  >
    <div
      className="container-fluid "
    >
      <div
        className="row align-items-center"
      >
        <div
          className="col-12 col-md-12 order-first"
        />
        <div
          className="col-12 col-md text-center"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`DefaultPage renders ContentBlock without background layout 1`] = `
<div
  className="container-fluid "
>
  <div
    className="row align-items-center"
  >
    <div
      className="col-12 col-md-12 order-first"
    />
    <div
      className="col-12 col-md text-center"
    >
      <h6 />
      <h2 />
      <div />
    </div>
  </div>
</div>
`;
